### PR TITLE
Markdown documentation

### PR DESCRIPTION
**Pull request by ferrarimarco**
Manual conversion from PDF to markdown, that give us the ability to do version control of the docs.

This PR also adds Jekyll to render the markdown into web pages, public via github pages here:
https://googlecloudplatform.github.io/solutions-viai-edge-provisioning-configuration/

### CODE REVIEW

File: .github/workflows/pages.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+# documentation.
+
+# Sample workflow for building and deploying a Jekyll site to GitHub Pages
+name: Deploy Jekyll site to Pages
+
+on:
+  push:
+    branches:
+      - "main"
+      - "markdown-documentation"
```
Comment (ferrarimarco) [inline .github/workflows/pages.yml:13]: You can probably remove this one.

File: config/lint/.markdown-lint.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+###############
+# Rules by id #
+###############
+
+MD013: false
+MD025: false
+MD029: false
+MD033: false
+MD034: false
+MD036: false
```
Comment (ferrarimarco) [inline config/lint/.markdown-lint.yml:27]: Why are we disabling these checks? The line length one is probably fine to disable, but we should perhaps fix the rest of the issues, unless there's a reason for not doing that (hard formatting, etc...).

File: docs/Gemfile [hunk 3]
```diff
@@ -0,0 +1,7 @@
+source 'https://rubygems.org'
```
Comment (ferrarimarco) [inline docs/Gemfile:1]: You need to watch this file with dependabot.

**Review by ferrarimarco — COMMENTED** [review]
A few small things to check.

File: .github/workflows/pages.yml [hunk 4]
```diff
@@ -0,0 +1,72 @@
+# This workflow uses actions that are not certified by GitHub.
+# They are provided by a third-party and are governed by
+# separate terms of service, privacy policy, and support
+# documentation.
+
+# Sample workflow for building and deploying a Jekyll site to GitHub Pages
```
Comment (ferrarimarco) [inline .github/workflows/pages.yml:6]: Remove.

File: .github/workflows/pages.yml [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+name: Deploy Jekyll site to Pages
```
Comment (ferrarimarco) [inline .github/workflows/pages.yml:7]: The process implemented in this workflow is not documented in any way.

Additionally, there's no way to run this outside GitHub Actions (unless you use more third-party tools, such as Act).

How can we evaluate the result of a change before actually merging a PR?

File: .github/workflows/pages.yml (hunk 4 above)
Comment (rbarberop) [reply .github/workflows/pages.yml]: Done

File: .github/workflows/pages.yml (hunk 5 above)
Comment (rbarberop) [reply .github/workflows/pages.yml:2]: Done

File: docs/Gemfile (hunk 3 above)
Comment (rbarberop) [reply docs/Gemfile:1]: Done, but please review...

**Review by ferrarimarco — COMMENTED** [review]
Some things to check.

File: .github/workflows/pages.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (rbarberop) [reply .github/workflows/pages.yml:8]: I'll do that as soon as we merge this PR, because if not, any changes in the docs folder won't be rendered.

I was planning on creating a tiny PR removing that branch after.

File: .github/workflows/pages.yml [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+on:
```
[the remark this answers is not in the document]
Comment (rbarberop) [reply .github/workflows/pages.yml:4]: The recommended way to do that, based on [GitHub pages + Jekyll](https://docs.github.com/en/pages/setting-up-a-github-pages-site-with-jekyll/testing-your-github-pages-site-locally-with-jekyll#building-your-site-locally) is to use Jekyll locally. I have documented it here: 

https://github.com/GoogleCloudPlatform/solutions-viai-edge-provisioning-configuration/blob/markdown-documentation/docs/development.md#render-the-documentation-as-static-html-using-jekyll-locally

File: config/lint/.markdown-lint.yml (hunk 2 above)
Comment (rbarberop) [reply config/lint/.markdown-lint.yml:31]: Done.

File: .github/workflows/pages.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ferrarimarco) [reply .github/workflows/pages.yml:2]: I saw your other comment. Can you please containerize that process, so that users don't have to install Jekyll? Thanks!

File: docs/development.md [hunk 7]
```diff
@@ -1,4 +1,4 @@
-# Develop VIAI Edge
+# Developing VIAI Edge
```
Comment (ferrarimarco) [inline docs/development.md:1]: Revert this change. We don't use gerunds on headings :)

File: docs/Gemfile (hunk 3 above)
Comment (ferrarimarco) [inline docs/Gemfile:1]: Commit the related `Gemfile.lock` to make the build more reproducible.

File: docs/development.md [hunk 8]
```diff
@@ -64,3 +64,15 @@ To run build container images locally, do the following:
...
 ```
+
+## Render the documentation as static HTML using Jekyll locally
+
+This repository uses Jekyll and GitHub actions to render the Markdown documentation
+under the `/docs` folder as static HTML.
+
+The configuration for the Jekyll site is at `docs/_config.yml` and the `docs/Gemfile` files.
+
+The configuration for the associated GitHub actions is at `.github/workflows/pages.yml`.
```
Comment (ferrarimarco) [inline docs/development.md:75]: Refactor

```md
GitHub actions is at `.github/workflows/pages.yml`
```

as:

```md
[GitHub Actions workflow](../.github/workflows/pages.yml)
```

File: edge-server/readme.md [hunk 9]
```diff
@@ -3,10 +3,8 @@ Add new Kubernetes runtime
```
Comment (ferrarimarco) [inline edge-server/readme.md:3]: Do we need this README now that we have the full instructions?

File: media/USB/readme.md [hunk 10]
```diff
@@ -3,10 +3,8 @@ Add new installation media type
```
Comment (ferrarimarco) [inline media/USB/readme.md:3]: Do we need this README now that we have the full instructions?

File: docker/documentation-site-builder/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    build-essential \
+    git \
+    && rm -rf /var/lib/apt/lists/*
+
+# used in the jekyll-server image, which is FROM this image
+COPY docker-entrypoint.sh /usr/local/bin/
+
+RUN gem update --system && gem install jekyll && gem cleanup
```
Comment (ferrarimarco) [inline docker/documentation-site-builder/Dockerfile:13]: We should use the `Gemfile` and the `Gemfile.lock` here. Otherwise, this will always install the latest version of Jekyll when building the image.

File: docker/documentation-site-builder/docker-entrypoint.sh [hunk 12]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+set -e
+
+if [ ! -f ../../docs/Gemfile ]; then
+  echo "ERROR: Gemfile not found"
+  exit 1
+fi
+
+bundle install --retry 5 --jobs 20
+
```
Comment (ferrarimarco) [inline docker/documentation-site-builder/docker-entrypoint.sh:10]: No need for this installation after you refactor the Dockerfile to install gems at build time.

File: docs/Gemfile [hunk 13]
```diff
@@ -0,0 +1,7 @@
...
+
+gem "jekyll", "~> 4.3.2" # installed by `gem jekyll`
+# gem "webrick"        # required when using Ruby >= 3 and Jekyll <= 4.2.2
+
+gem "just-the-docs", "0.7.0" # pinned to the current release
+# gem "just-the-docs"        # always download the latest release
```
Comment (ferrarimarco) [inline docs/Gemfile:7]: Remove commented gems.

File: docs/Gemfile (hunk 3 above)
Comment (rbarberop) [reply docs/Gemfile:1]: Done

File: docker/documentation-site-builder/Dockerfile (hunk 11 above)
Comment (rbarberop) [reply docker/documentation-site-builder/Dockerfile]: Done

File: docker/documentation-site-builder/docker-entrypoint.sh (hunk 12 above)
Comment (rbarberop) [reply docker/documentation-site-builder/docker-entrypoint.sh]: Eliminated

File: docs/Gemfile (hunk 13 above)
Comment (rbarberop) [reply docs/Gemfile]: Done

File: docker/documentation-site-builder/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    git \
+    && rm -rf /var/lib/apt/lists/*
+
+WORKDIR /docs
+
+# First copy the Gemfile
+COPY docs/Gemfile /docs/
+COPY docs/Gemfile.lock /docs/
+
+RUN gem update --system && gem install jekyll && gem install just-the-docs && gem cleanup
```
Comment (ferrarimarco) [inline docker/documentation-site-builder/Dockerfile:16]: A few things here:

1- Don't update all gems because this makes the build not reproducible.
2- Install gems with `bundle install`, so you pick the contents from the `Gemfile`.

File: docker/documentation-site-builder/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+# Copying the docs last, to maximize docker cache 
+COPY docs/ /docs/
+
+EXPOSE 4000
+
+# CMD [ "bundle", "exec", "jekyll", "serve", "--force_polling", "-H", "0.0.0.0", "-P", "4000" ]
```
Comment (ferrarimarco) [inline docker/documentation-site-builder/Dockerfile:23]: Remove this commented line.

File: docker/documentation-site-builder/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+CMD ["bundle", "exec", "jekyll", "serve", "-H" , "0.0.0.0", "-P", "4000" ]
```
Comment (ferrarimarco) [inline docker/documentation-site-builder/Dockerfile:24]: Split this in two:

`ENTRYPOINT`: `bundle exec jekyll`
`CMD`: the rest.

File: tests/build-documentation-site.sh [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+. ./scripts/common.sh
+
+echo "Building Jekyll"
+
+docker build -t jekyll:4.3.2 -f docker/documentation-site-builder/Dockerfile .
+
+echo "Rendering documentation site"
+echo "Open your browser at http://127.0.0.1:4000/docs/ to review the documentation locally"
+
+docker run -p 4000:4000 jekyll:4.3.2
```
Comment (ferrarimarco) [inline tests/build-documentation-site.sh:30]: Refactor `jekyll:4.3.2` -> `viai-edge-documentation-site:latest`

**Review by ferrarimarco — COMMENTED** [review]
A couple of things to check. We're almost there!

File: docker/documentation-site-builder/Dockerfile (hunk 14 above)
Comment (rbarberop) [reply docker/documentation-site-builder/Dockerfile]: Done

File: docker/documentation-site-builder/Dockerfile (hunk 16 above)
Comment (rbarberop) [reply docker/documentation-site-builder/Dockerfile]: Done

File: docker/documentation-site-builder/Dockerfile (hunk 15 above)
Comment (rbarberop) [reply docker/documentation-site-builder/Dockerfile]: Done

File: tests/build-documentation-site.sh (hunk 17 above)
Comment (rbarberop) [reply tests/build-documentation-site.sh]: Done

File: docker/documentation-site-builder/Dockerfile [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+SHELL ["/bin/bash", "-eo", "pipefail", "-c"]
+
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    build-essential \
+    git \
+    && rm -rf /var/lib/apt/lists/*
+
+WORKDIR /docs
+
+# First copy the Gemfile
```
Comment (ferrarimarco) [inline docker/documentation-site-builder/Dockerfile:12]: We need the `Gemfile.lock` to make the build reproducible.

File: docs/deployment/provisionviaiapp.md [hunk 19]
```diff
@@ -0,0 +1,7 @@
+# Provisioning VIAI Edge app
```
Comment (ferrarimarco) [inline docs/deployment/provisionviaiapp.md:1]: Just noticed this one. Is this document something to elaborate?

File: .github/workflows/pages.yml [hunk 20]
```diff
@@ -0,0 +1,82 @@
+# Workflow for building and deploying a the /docs folder as static Jekyll site to GitHub Pages
+name: Deploy Jekyll site to Pages
+
+on:
```
Comment (ferrarimarco) [inline .github/workflows/pages.yml:4]: Can you add the `pull_request` event here, so the jobs listed below (minus the `deploy` one, see my other comment below) run on against PRs as well?

File: .github/workflows/pages.yml [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+  push:
+    branches:
+      - "main"
+      - "markdown-documentation"
```
Comment (ferrarimarco) [inline .github/workflows/pages.yml:8]: Remember that when you're going to merge this PR, this branch is going to be deleted, so you can probably remove it now.

File: .github/workflows/pages.yml [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+    paths:
```
Comment (ferrarimarco) [inline .github/workflows/pages.yml:9]: Add the following paths to this list:

- This workflow: `.github/workflows/pages.yml`
- Container image descriptor directory: `docker/document-site-builder/**`
- Build script: `tests/build-documentation-site.sh`
- Common script: `scripts/common.sh`

If you don't want to be that granular, you can remove this list altogether.

File: .github/workflows/pages.yml [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+          working-directory: '${{ github.workspace }}/docs'
+      - name: Setup Pages
+        id: pages
+        uses: actions/configure-pages@v3
+      - name: Build with Jekyll
+        # Outputs to the './_site' directory by default
+        run: bundle exec jekyll build --baseurl "${{ steps.pages.outputs.base_path }}"
+        env:
+          JEKYLL_ENV: production
+      - name: Upload artifact
```
Comment (ferrarimarco) [inline .github/workflows/pages.yml:51]: (follow up to my comment above about triggering this workflow from PRs as well)

Add a condition to this step `if: ...` to run only if (this is an AND):

- We're in this repository
- This is a push event
- On the default branch

File: .github/workflows/pages.yml [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+        # Automatically uploads an artifact from the './_site' directory by default
+        uses: actions/upload-pages-artifact@v1
+        with:
+          path: "docs/_site/"
+
+  # Deployment job
+  deploy:
```
Comment (ferrarimarco) [inline .github/workflows/pages.yml:58]: (follow up to my comment above about triggering this workflow from PRs as well)

Add a condition to this step `if: ...` to run only if (this is an AND):

- We're in this repository
- This is a push event
- On the default branch

File: .github/workflows/pages.yml (hunk 21 above)
Comment (ferrarimarco) [reply .github/workflows/pages.yml:8]: Also, remove it from the allowed branches in the environment protection rules.

**Review by ferrarimarco — COMMENTED** [review]
HI @rbarberop !

A few minor things to check

File: docs/deployment/provisionviaiapp.md (hunk 19 above)
Comment (rbarberop) [reply docs/deployment/provisionviaiapp.md:1]: This is a residue from the original PDF instructions. The content is already covered in the other pages. Deleting.

File: .github/workflows/pages.yml (hunk 24 above)
Comment (rbarberop) [reply .github/workflows/pages.yml:59]: Done

File: .github/workflows/pages.yml (hunk 23 above)
Comment (rbarberop) [reply .github/workflows/pages.yml:51]: Done

File: .github/workflows/pages.yml (hunk 22 above)
Comment (rbarberop) [reply .github/workflows/pages.yml]: Done

File: .github/workflows/pages.yml (hunk 21 above)
Comment (rbarberop) [reply .github/workflows/pages.yml]: Done

File: .github/workflows/pages.yml (hunk 20 above)
Comment (rbarberop) [reply .github/workflows/pages.yml:4]: Done

File: .github/workflows/pages.yml [hunk 25]
```diff
@@ -2,12 +2,12 @@
 name: Deploy Jekyll site to Pages
 
 on:
+  pull_request:
+    branches:
+      - "main"
```
Comment (ferrarimarco) [inline .github/workflows/pages.yml:7]: Remove this. We want to run this workflow on every PR.

File: .github/workflows/pages.yml [hunk 26]
```diff
@@ -2,12 +2,12 @@
...
   push:
     branches:
       - "main"
```
Comment (ferrarimarco) [inline .github/workflows/pages.yml:10]: Remove this, we don't need to filter on the main branch only.

File: .github/workflows/pages.yml [hunk 27]
```diff
@@ -49,13 +49,15 @@ jobs:
         env:
           JEKYLL_ENV: production
       - name: Upload artifact
+        if: ${{ github.event_name == 'push' && github.repository == 'GoogleCloudPlatform/solutions-viai-edge-provisioning-configuration' && github.ref_name == 'default'}}
```
Comment (ferrarimarco) [inline .github/workflows/pages.yml:52]: this will match branches named `default`. We should probably use: `github.event.repository.default_branch` (without `'`)

File: .github/workflows/pages.yml [hunk 28]
```diff
@@ -49,13 +49,15 @@ jobs:
...
         # Automatically uploads an artifact from the './_site' directory by default
         uses: actions/upload-pages-artifact@v1
         with:
           path: "docs/_site/"
 
   # Deployment job
   deploy:
+    if: ${{ github.event_name == 'push' && github.repository == 'GoogleCloudPlatform/solutions-viai-edge-provisioning-configuration' && github.ref_name == 'default'}}
```
Comment (ferrarimarco) [inline .github/workflows/pages.yml:60]: this will match branches named `default`. We should probably use: `github.event.repository.default_branch` (without `'`)

**Review by ferrarimarco — COMMENTED** [review]
Small things.

File: .github/workflows/pages.yml (hunk 27 above)
Comment (rbarberop) [reply .github/workflows/pages.yml]: Done

File: .github/workflows/pages.yml (hunk 28 above)
Comment (rbarberop) [reply .github/workflows/pages.yml]: Done

File: .github/workflows/pages.yml (hunk 25 above)
Comment (rbarberop) [reply .github/workflows/pages.yml]: Done

File: .github/workflows/pages.yml (hunk 26 above)
Comment (rbarberop) [reply .github/workflows/pages.yml]: Done

**Comment by ferrarimarco** [comment]
LGTM. Merging this one forcefully because I created this PR.
